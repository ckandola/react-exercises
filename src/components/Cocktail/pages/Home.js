import React from 'react';
import CocktailList from '../CocktailList';
import SearchForm from '../SearchForm';

const Home = () => {
    return (
        <div>
            <h2>Home Page</h2>
        </div>
    );
};

export default Home;
