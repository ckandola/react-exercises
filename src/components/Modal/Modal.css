/*
=============== 
Variables
===============
*/

:root {
    --clr-white: #fff;
    --clr-red-dark: hsl(360, 67%, 44%);
    --clr-black: #222;
    --transition: all 0.3s linear;
    --spacing: 0.1rem;
    --radius: 0.25rem;
    --fixed-width: 620px;
  }

.modal-btn {
    text-transform: uppercase;
    background: var(--clr-black);
    color: var(--clr-white);
    padding: 0.375rem 0.75rem;
    letter-spacing: var(--spacing);
    display: inline-block;
    transition: var(--transition);
    font-size: 0.875rem;
    border: 2px solid var(--clr-black);
    cursor: pointer;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);
    border-radius: var(--radius);
    margin: 0.5rem;
  }
  .modal-btn:hover {
    color: var(--clr-black);
    background: transparent;
  }
  

.modal-main {
    min-height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
  }

.modal-overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.5);
    display: grid;
    place-items: center;
    transition: var(--transition);
    visibility: hidden;
    z-index: -1;
  }
  /* OPEN/CLOSE MODAL */
  .show-modal {
    visibility: visible;
    z-index: 10;
  }
  .modal-container {
    background: var(--clr-white);
    border-radius: var(--radius);
    padding-right: 2rem;
    width: 90vw;
    max-height: fit-content;
    min-height: 50vh;
    max-width: var(--fixed-width);
    text-align: center;
    display: grid;
    place-items: center;
    position: relative;
  }
  .close-modal-btn {
    position: absolute;
    top: 1rem;
    right: 1rem;
    font-size: 2rem;
    background: transparent;
    border-color: transparent;
    color: var(--clr-red-dark);
    cursor: pointer;
  }
