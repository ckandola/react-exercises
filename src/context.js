import React, { useState, useContext, useReducer, useEffect, useCallback } from 'react';
import {default as smenulinks} from './components/StripeMenu/data';

import cartItems from './components/Cart/data';
import {default as cartReducer} from './components/Cart/reducer';

const AppContext = React.createContext();
    // gives both Provider and Consumer

const AppProvider = ({children}) => {
    const [isModalOpen, setIsModalOpen] = useState(false);
    
    // StripeMenu ---

    const [isSidebarOpen, setIsSidebarOpen] = useState(false);
    const [isSubmenuOpen, setIsSubmenuOpen] = useState(false);
    const [location, setLocation] = useState({});
    const [page, setPage] = useState({page: '', links:[]});

    const openSidebar = () => {
        setIsSidebarOpen(true);
    }
    
    const closeSidebar = () => {
        setIsSidebarOpen(false);
    }
    
    const openSubmenu = (text, coordinates) => {
        setPage(smenulinks.find(x => x.page === text));
        setLocation(coordinates);
        setIsSubmenuOpen(true);
    }
    
    const closeSubmenu = () => {
        setIsSubmenuOpen(false);
    }

    // Modal ---

    const openModal = () => {
        setIsModalOpen(true);
    }

    const closeModal = () => {
        setIsModalOpen(false);
    }

    // Cart ---
    const cartUrl = 'https://course-api.com/react-useReducer-cart-project';
    const initialCartState = {
        cartLoading: false,
        cart: cartItems,
        cartTotal: 0,
        cartAmount: 0
    };
    const [cartState, cartDispatch] = useReducer(cartReducer, initialCartState);
    const clearCart = () => {
        cartDispatch({type: 'CLEAR_CART'});
    };
    const removeCartItem = id => {
        cartDispatch({type: 'REMOVE_ITEM', payload: id});
    }
    const increaseCartAmt = id => {
        cartDispatch({type: 'INCREASE', payload: id});
    }
    const decreaseCartAmt = id => {
        cartDispatch({type: 'DECREASE', payload: id});
    }
    const fetchCartData = async () => {
        cartDispatch({type: 'LOADING'});
        const response = await fetch(cartUrl);
        const cart = await response.json();
        cartDispatch({type: 'DISPLAY_ITEMS', payload: cart});
    }

    useEffect(() => {
        cartDispatch({type: 'GET_TOTALS'});
    }, [cartState.cart]);

    useEffect(() => {
        fetchCartData();
    }, []);

    // CocktailDB ---
    const cocktailUrl = 'https://www.thecocktaildb.com/api/json/v1/1/search.php?s=';
    const [cocktailLoading, setCocktailLoading] = useState(true);
    const [cocktailSearchTerm, setCocktailSearchTerm] = useState('a');
    const [cocktails, setCocktails] = useState([]);

    const fetchDrinks = useCallback(async () => {
        try {
            setCocktailLoading(true);
            const response = await fetch(`${cocktailUrl}${cocktailSearchTerm}`);
            const data = await response.json();
            const {drinks} = data;
            if (!drinks) {
                setCocktails([]);
            } else {
                const newDrinks = drinks.map(item => {
                    const {idDrink, strDrink, strDrinkThumb, strAlcoholic, strGlass} = item;
                    return {id: idDrink, name: strDrink, img: strDrinkThumb, isAlcoholic: strAlcoholic, glass: strGlass};
                });
                setCocktails(newDrinks);
            }
            setCocktailLoading(false);
        } catch (error) {
            console.error(error);
            setCocktailLoading(false);
        }
    }, [cocktailSearchTerm]);

    useEffect(() => {
        fetchDrinks();
    }, [fetchDrinks, cocktailSearchTerm]);

    return (
        <AppContext.Provider value={{
            isModalOpen, openModal, closeModal,
            
            isSidebarOpen, openSidebar, closeSidebar,
            isSubmenuOpen, openSubmenu, closeSubmenu,
            location, page,

            ...cartState, clearCart, removeCartItem, 
            increaseCartAmt, decreaseCartAmt, fetchCartData,

            cocktailLoading, cocktails, setCocktailSearchTerm
        }}>
            {children}
        </AppContext.Provider>
    );
};

// custom hook
export const useGlobalContext = () => {
    return useContext(AppContext);
};

export {AppContext, AppProvider};
