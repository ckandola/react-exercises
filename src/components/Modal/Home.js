import React from 'react';
import {useGlobalContext} from '../../context';
import './Modal.css';

const Home = () => {
    const {openModal} = useGlobalContext();

    return (
        <main className="modal-main">
            <button 
                className="modal-btn"
                onClick={openModal}>
                Show Modal
            </button>
        </main>
    );
};

export default Home;
