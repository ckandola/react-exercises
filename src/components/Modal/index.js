import React from 'react';
import {FaTimes} from 'react-icons/fa';
import {useGlobalContext} from '../../context';
import './Modal.css';
import PropTypes from 'prop-types';

const Modal = ({children}) => {
    const {isModalOpen, closeModal} = useGlobalContext();
    return (
        <div className={`modal-overlay ${isModalOpen && 'show-modal'}`}>
            <div className="modal-container">
                {children}
                <button 
                    className="close-modal-btn"
                    onClick={closeModal}>
                    <FaTimes />
                </button>
            </div>
        </div>
    );
};

export default Modal;

Modal.propTypes = {
    children: PropTypes.node
};
