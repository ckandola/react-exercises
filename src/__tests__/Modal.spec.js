import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Home from '../components/Modal/Home';
import Modal from '../components/Modal';
import * as AppProvider from '../context';

describe('Modal tests', () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(
            <div>
                <Home />
                <Modal>test1</Modal>
            </div>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('opens the modal when Show Modal is clicked', () => {
        const open = jest.fn();
        const contextValues = { isModalOpen: false, closeModal: () => {}, openModal: open };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(
            <div>
                <Home />
                <Modal>test2</Modal>
            </div>
        );
        wrapper.find('.modal-btn').simulate('click');
        expect(open.mock.calls.length).toBe(1);
    });

    it('closes the modal when the red X is clicked', () => {
        const close = jest.fn();
        const contextValues = { isModalOpen: true, closeModal: close, openModal: () => {} };
        jest.spyOn(AppProvider, 'useGlobalContext').mockImplementation(() => contextValues);
        const wrapper = mount(
            <div>
                <Home />
                <Modal>test3</Modal>
            </div>
        );
        expect(wrapper.find('.show-modal')).toHaveLength(1);
        wrapper.find('.close-modal-btn').simulate('click');
        expect(close.mock.calls.length).toBe(1);
    });
});
