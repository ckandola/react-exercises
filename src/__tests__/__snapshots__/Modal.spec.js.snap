// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal tests matches the snapshot 1`] = `
<div>
  <Home />
  <Modal>
    test1
  </Modal>
</div>
`;
